pkg/controllers/provisioning/scheduling/theoreticalnode.go

```diff
@@ -0,0 +1,137 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"errors"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// TheoreticalNode is a set of constraints, compatible pods, and possible instance types that could fulfill
+// these constraints. This will be turned into one or more actual node instances within the cluster after bin packing.
+type TheoreticalNode struct {
```

Comment:
Optional, I think we could reasonably just call this `Node`, which would make it a `scheduling.Node` instead of a `v1.Node`. We're already calling the variable names `node` anyways. I'm a big fan of context driven naming, where the names themselves are very simple, and any contextual adjectives are replaced by the package's context. This works even better with natural package names. It's not a perfect fit like `simulation.Node` or `simulated.Node`, but good enough, I think.